tests/Model/BSONIteratorTest.php

```diff
@@ -0,0 +1,133 @@
+<?php
+
+namespace MongoDB\Tests\Model;
+
+use MongoDB\Exception\UnexpectedValueException;
+use MongoDB\Model\BSONArray;
+use MongoDB\Model\BSONDocument;
+use MongoDB\Model\BSONIterator;
+use MongoDB\Tests\TestCase;
+use stdClass;
+
+class BSONIteratorTest extends TestCase
+{
+    /**
+     * @dataProvider provideTypeMapOptionsAndExpectedDocuments
+     */
+    public function testValidValues(array $typeMap = null, $binaryString, array $expectedDocuments)
+    {
+        $bsonIt = new BSONIterator($binaryString, ['typeMap' => $typeMap]);
+
+        $results = iterator_to_array($bsonIt);
+
+        $this->assertSameDocuments($expectedDocuments, $results);
```

Comment:
This should use `assertEquals()` to ensure we're checking expected types after type map conversion (as is done in AggregateFunctionalTest).